src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,127 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+struct Crystal::ZeroOneOrMany(T)
+  include Indexable(T)
+
+  getter value : Nil | T | Array(T)
+
+  def initialize
+    @value = nil
+  end
+
+  def initialize(@value : T)
+  end
+
+  protected def initialize(values : Array(T))
+    @value =
+      case values.size
+      when 0
+        nil
+      when 1
+        values.first
+      else
+        values
+      end
+  end
+
+  def unsafe_fetch(index : Int)
+    value = @value
+    case value
+    in Nil
+      raise "BUG: called ZeroOneOrMany#unsafe_fetch but value is nil"
+    in T
+      value
```

Comment:
I guess that makes sense, if for some reason someone calls this method directly.

Reply:
Done!

Reply:
Since this is implementing `Indexable#[]`, wouldn't `IndexError` make more sense? Same for `Nil` branch.

Reply:
Sounds good... I mean, this is internal for Crystal and it's likely to never happen. We are just making that code more complex (and likely slower) for no real reason. But okay...

Reply:
Done!

Reply:
I don't think `IndexError` impacts runtime performance?
The index check itself certainly does, but it's only a simple comparison. I think it should be worth it to avoid potential bugs in the future.